## Exercise 4.6 Combining & Exporting Data

### This script contains:

### 1. Import all datasets

### 2. Checking the datasets

### 3. Determine a way to combine the dataset

### 4. Confirm the results of the merge using the merge flag

### 5. Export the data

In [2]:
# Import libraries

import pandas as pd
import numpy as np
import os

### 1. Import all datasets

In [3]:
# Creating path and Importing datasets

path = r'/Users/queency_yustiawan/Study/CF/data analyst course/immersion/A4 exercise/Instacart Basket Analysis'

df_ords_prods_combined = pd.read_pickle (os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

df_prods = pd.read_csv (os.path.join (path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

### 2. Checking the datasets

In [4]:
# Checking data output

df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   order_interval_days  product_id  add_to_cart_order  reordered _merge  
0                  NaN         196                  1          0   both  
1                  NaN       14084                  2          0   both  
2                  NaN       12427                  3          0   both  
3                  NaN       26088                  4          0   both  
4                  NaN       26405                  5          0   both

In [5]:
df_ords_prods_combined.shape

(32434489, 10)

In [6]:
# Check the output

df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [9]:
# Removing unnamed column

df_prods = df_prods.drop(columns = ['Unnamed: 0'])

In [10]:
df_prods.shape

(49672, 5)

### 3. Determine a way to combine the dataset

#### To combine both datasets, the best way is to merge them, as they dont match in shape and they have a common key "product_id". 

In [12]:
# Merge data

df_ords_prods_merge = df_ords_prods_combined.merge(df_prods, on = 'product_id', indicator = True)

ValueError: Cannot use name of an existing column for indicator column

#### The merge was failed due to the df_ords_prods_combined data already has the identifier column of '_merge'. So we will removed the '_merge' column first.

In [15]:
# Removing '_merge' column

df_ords_prods_combined = df_ords_prods_combined.drop(columns = ['_merge'])

In [16]:
# checking output

df_ords_prods_combined.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   order_interval_days  product_id  add_to_cart_order  reordered  
0                  NaN         196                  1          0  
1                  NaN       14084                  2          0  
2                  NaN       12427                  3          0  
3                  NaN       26088                  4          0  
4                  NaN       26405                  5          0

In [17]:
# Merge data

df_ords_prods_merge = df_ords_prods_combined.merge(df_prods, on = 'product_id', indicator = True)

In [18]:
# Check output

df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   order_interval_days  product_id  add_to_cart_order  reordered product_name  \
0                  NaN         196                  1          0         Soda   
1                 15.0         196                  1          1         Soda   
2                 21.0         196                  1          1         Soda   
3                 29.0         196                  1          1         Soda   
4                 28.0         196                  1          1         Soda   

   aisle_id  department_id  prices _merge  
0        77              7     9.0   both  
1        77              7     9.0   both  
2        77              7     9.0   both  
3        77              7     9.0   both  
4        77              7     9.0   both

### 4. Confirm the results of the merge using the merge flag

In [19]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

#### The merge results in 32,404,859 rows in which they have data from both datasets since we used inner join

### 5. Export the dataset

In [1]:
# Export data to pkl

df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merge.pkl'))

NameError: name 'df_ords_prods_merge' is not defined